File: nim/day03.nim
import sequtils

func traverse(grid: seq[string], right, down: int): int =
  var x: int
  for y in countup(0, grid.high, down):
    if grid[y][x] == '#': inc result
    x = (x + right) mod grid[0].len

func part2(grid: seq[string]): int =
  result = traverse(grid, 1, 2)
  for right in [1, 3, 5, 7]:
    result *= traverse(grid, right, 1)

let grid = toSeq("inputs/03.txt".lines)
echo traverse(grid, 3, 1)
echo part2(grid)

File: python/day03.py
from itertools import count
from operator import mul
from functools import reduce

grid = [line.strip() for line in open("inputs/03.txt").readlines()]

traverse = lambda right, down: sum(grid[y][x % len(grid[0])] == '#'
                                   for (x, y) in zip(count(0, right),
                                                     range(0, len(grid), down)))

print(traverse(3, 1))

slopes = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
print(reduce(mul, (traverse(*s) for s in slopes)))
